### DISCUSSION TITLE
Proposal: Default to borderless when tabbing out of fullscreen then switching back like stable.

### DISCUSSION THREAD

**Discussion by TheFantasticLoki** [ppy/osu#17172] [Ideas] [answered]
Pretty self explanatory but I really miss this feature especially for multiplayer cause I often do stuff on my second monitor when waiting in lobby and can't see when everyone is ready.
I yes I did a search for borderless and found nothing.

**Comment by Walavouchey** [comment]
~~Is this just about being able to move the cursor to another monitor while in fullscreen? You can hit F11 to cycle between display modes, and although it's a bit more inconvenient than in stable, I don't think alt-tabbing should switch display modes~~

I think I misunderstood. Stable lets you move the cursor off-screen and I presume seamlessly switches to borderless when clicking something else in order to let the game be visible, which also happens when alt-tabbing. Tbh it's a useful feature

**Reply by bdach** [reply]
f11 or alt-enter.

also note that exclusive fullscreen is better for gameplay than borderless due to latency concerns so this may not be desired by all users.

also i don't know what "switching back like stable" refers to. i'm pretty sure stable doesn't change borderless to fullscreen when you alt-tab.

**Reply by Walavouchey** [reply]
I'm not entirely sure how stable does it, but stable lets you alt-tab out of the game and move your cursor (with the right cursor confine setting) to other monitors while in fullscreen mode, which I think is achieved by the method I just described. When the user alt-tabs or clicks back into the game, it switches back from the temporary borderless mode to fullscreen. Have a try yourself

The thing is that stable is presumably always in fullscreen (with the fullscreen setting) as long as it's focused, so I see no issue with latency there

**Reply by peppy** [reply]
I'm not sure we want to do this again, but we may have to. The reason stable likely works like this is because we're doing some very weird (and potentially dated) hacks to make sure exclusive fullscreen works across all GPU drivers (yes, it's up to the driver to enter exclusive mode in OpenGL).

The difference in behaviour here also likely the reason exclusive fullscreen is not working on some hardware (optimus/dualgpu laptops for example).

**Reply by peppy** [reply]
Recent SDL change may be relevant: https://github.com/libsdl-org/SDL/pull/5389

**Comment by respektive** [comment]
I second this! This has been bugging me for quite a bit now. 

I will add though, that even when I am running borderless mode the window still minimises when I alt-tab or click a window on another screen.

**Comment by peppy** [comment]
May be able to implement this with https://github.com/libsdl-org/SDL/pull/5537

**Reply by TheFantasticLoki** [reply]
Has this been considered for implementation any time soon?

**Comment by april83c** [comment]
Any updates on this? I'm a streamer and I use text-to-speech because I can't use my voice, which requires me to tab out whenever I want to talk.

With the current Lazer behavior, this causes osu! to visually freeze on the stream while I'm tabbed out, but with Stable (and the option pictured below disabled), it doesn't.

![image](https://github.com/ppy/osu/assets/124521665/f314310b-3cbf-470e-ae75-ed7aa356cc79)

**Comment by Susko3** [comment, marked as answer]
SDL seems to natively allow this with [`SDL_HINT_VIDEO_MINIMIZE_ON_FOCUS_LOSS`](https://wiki.libsdl.org/SDL2/SDL_HINT_VIDEO_MINIMIZE_ON_FOCUS_LOSS) [which we currently set to `false`](https://github.com/ppy/osu-framework/blob/fffe2e749195d8159aeeebf4d2fca23f991329c6/osu.Framework/Platform/SDL2Window.cs#L233).

Since application-set SDL hints can be overridden by environment variables, you can [set the environment variables](https://phoenixnap.com/kb/windows-set-environment-variable#ftoc-heading-4) like outlined below to get the result you want:
| Name | Value |
|--------|--------|
| `SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS` | `0` | 

Please note that this will apply to all SDL/SDL2 applications you run.

---

We could easily expose this as a setting. I would also expect the cursor to not be confined if clicking outside the window won't minimise it.
